cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/config_renderer.py

```diff
@@ -0,0 +1,177 @@
+            compute_resource_config, self.instance_types, instance_types_data, memory_ratio
+        )
+        self.EFA_enabled, self.EFA_gdr_support = _get_efa_settings(compute_resource_config)
+        self.instance_types = _get_instance_types(compute_resource_config)
+        self.vcpus_count, self.threads_per_core = _get_min_vcpus(self.instance_types, instance_types_data)
+        self.gpu_count, self.gpu_type = _get_min_gpu_count_and_type(self.instance_types, instance_types_data, log)
+
+    def static_node_name(self):
+        """Render the NodeName section for static nodes."""
+        return self._node_name("st", self.static_nodes)
+
+    def dynamic_node_name(self):
+        """Render the NodeName section for dynamic nodes."""
+        return self._node_name("dy", self.dynamic_nodes)
+
+    def vcpus(self) -> int:
+        """Return the number of vcpus according to disable_hyperthreading and instance features."""
+        return (
+            self.vcpus_count
+            if not self.disable_simultaneous_multithreading
+            else (self.vcpus_count // self.threads_per_core)
+        )
+
+    def gpus(self) -> dict:
+        """Return the number of GPUs and type for the compute resource."""
+        return {"count": self.gpu_count, "type": self.gpu_type}
+
+    def render(self):
+        """Launch the rendering process."""
+        config = ""
+        if self.static_nodes > 0:
+            config += f"NodeName={self.static_node_name()}{ self._definitions() }\n"
+        if self.dynamic_nodes > 0:
+            config += f"NodeName={self.dynamic_node_name()}{ self._definitions(dynamic=True) }\n"
+
+        return config
+
+    def as_node_set(self):
+        """Alternative rendering for the NodeSet definition."""
+        node_set = ""
+        if self.static_nodes > 0:
+            node_set += f"{self.static_node_name()},"
+        if self.dynamic_nodes > 0:
+            node_set += f"{self.dynamic_node_name()},"
+
+        return node_set
+
+    def as_gres(self):
+        """Alternative rendering for the gres config."""
+        gres_render = ""
+        if self.gpu_count > 0:
+            if self.static_nodes > 0:
+                gres_render += (
+                    f"NodeName={self.static_node_name()} Name=gpu "
+                    f"Type={ self.gpu_type } File=/dev/nvidia[0-{ self.gpu_count - 1 }]\n"
+                )
+            if self.dynamic_nodes > 0:
+                gres_render += (
+                    f"NodeName={self.dynamic_node_name()} Name=gpu "
+                    f"Type={ self.gpu_type } File=/dev/nvidia[0-{ self.gpu_count - 1 }]\n"
+                )
+
+        return gres_render
+
+    def _definitions(self, dynamic=False):
+        definitions = f" CPUs={ self.vcpus() } RealMemory={ self.real_memory } State=CLOUD {self._features(dynamic)}"
+
+        if self.has_gpu and self.gpu_count > 0:
+            definitions += f" Gres=gpu:{ self.gpu_type }:{self.gpu_count}"
```

Comment:
Nitpick: the spacing between for the `{ }` is a bit inconsistent throughout the file (I'm surprised the linters are not complaining about it).